doc/Container Hardening/SONiC_container_hardening_HLD.md

```diff
@@ -0,0 +1,349 @@
+1. Privileges
+2. Network
+3. Capabilities
+4. Mount namespace
+5. Cgroups
+6. Etc
+
+For now, we will focus on #1 & #2
+
+Further guidelines and requirements will be brought upon in the future on-demand.
+
+## 3. Architecture Design
+
+### 3.1 Root privileges
+
+When removing the root privileges from a specific container - we are required to remove the `--privileged` flag and add the required missing Linux capabilities to the docker, or alternatively adjust the container so that it does not require root privileges to perform any action.
+
+### 3.2 net=host
+
+Removing the `net=host` is required to prevent the container from accessing the full network scope of the host and system.
+When doing this removal - we will start getting failures from devices that require external access and packet transfers between the container and the host to the interfaces.
+In order to overcome this obstacle - we have a few options here:
+- using `--net=bridge` and port forwarding
+
+## 4. High-Level Design
+
+### 4.1 Root privileges removal
+Removing the `--privileged` flag is done by editing the docker_image_ctl.j2 file:
+
+docker_image_ctl.j2 file
+
+	docker create {{docker_image_run_opt}} \ # *Need to modify this parameter "docker_image_run_opt" to not contain the --privileged flag*
+	{%- if docker_container_name != "database" %}
+		--net=$NET \
+		--uts=host \{# W/A: this should be set per-docker, for those dockers which really need host's UTS namespace #}
+	{%- endif %}
+	{%- if docker_container_name == "database" %}
+		-p 6379:6379 \
+	{%- endif %}
+		-e RUNTIME_OWNER=local \
+	{%- if install_debug_image == "y" %}
+		-v /src:/src:ro -v /debug:/debug:rw \
+	{%- endif %}
+	{%- if '--log-driver=json-file' in docker_image_run_opt or '--log-driver' not in docker_image_run_opt %}
+		--log-opt max-size=2M --log-opt max-file=5 \
+	{%- endif %}
+
+This will cause the docker file to be altered in the following manner:
+
+**database.sh file**
+
+	docker create --privileged -t -v /etc/sonic:/etc/sonic:ro \ # *Need to remove the --privileged flag*
+		-p 6379:6379 \
+		-e RUNTIME_OWNER=local \
+		--log-opt max-size=2M --log-opt max-file=5 \
+		--tmpfs /tmp \
+		$DB_OPT \
+		$REDIS_MNT \
+		-v /usr/share/sonic/device/$PLATFORM:/usr/share/sonic/platform:ro \
+		--tmpfs /var/tmp \
+		--env "NAMESPACE_ID"="$DEV" \
+		--env "NAMESPACE_PREFIX"="$NAMESPACE_PREFIX" \
+		--env "NAMESPACE_COUNT"=$NUM_ASIC \
+		--name=$DOCKERNAME \
+		docker-database:latest \
+		|| {
+			echo "Failed to docker run" >&1
+			exit 4
+		}
+
+#### Docker privileges
+Removing the root privileges from the docker container - will remove some Linux capabilities that are inherited from the root level permissions.
+
+Running the capabilities list command on a privileged container, this includes all capabilities captured in both [Table 3: Default Linux capabilities](#table-3-default-linux-capabilities) and [Table 4: Extended Linux capabilities](#table-4-extended-linux-capabilities)
```

Comment:
Let's add a section to list out a table what containers can be hardened or not?

Reply:
We can add a section.
Contemplating if this should be in the HLD or in a seperate discussion

Reply:
Done